### PR TITLE
nixos/switch-to-configuration: Fix backslashes in unit names

### LINKED ISSUE #160445: `Failed to open` systemd service file warning in `switch-to-configuration`
### Describe the bug

Switching NIxOS configurations when using `networking.wireguard.interfaces` produces the following warning:
```
Failed to open /etc/systemd/system/wireguard-wg0-peer-Ks9y…BIx3d.service: No such file or directory at /run/current-system/specialisation/spec/bin/switch-to-configuration line 137.
Failed to open /nix/store/6m1ckayw2n6i7aqbzqd9qr7cm6iwp66h-nixos-system-machine-22.05.git.86a0c32d907/etc/systemd/system/wireguard-wg0-peer-Ks9y…SBIx3d.service: No such file or directory at /run/current-system/specialisation/spec/bin/switch-to-configuration line 137.
```

### Steps To Reproduce

1. Checkout `nixpkgs` b2de46276d01a88d068aa9b04bba387d2e7c991d (current master)
2. Create `switch-to-configuration-test.nix` file in the `nixpkgs` directory with the following content:
```nix
import ./nixos/tests/make-test-python.nix {
  name = "switch-to-configuration";
  machine = let keys = import ./nixos/tests/wireguard/snakeoil-keys.nix; in
    {
      networking.wireguard.interfaces.wg0 = {
        inherit (keys.peer0) privateKey;

        peers = [{ allowedIPs = [ "0.0.0.0/0" ]; inherit (keys.peer1) publicKey; }];
      };

      specialisation.spec.configuration = { };
    };

  testScript = ''
    machine.wait_for_unit("multi-user.target")
    out = machine.succeed("/run/current-system/specialisation/spec/bin/switch-to-configuration test 2>&1 | tee -a /dev/stderr")
    assert 'Failed to open' not in out
  '';
}
```
3. Run `nix build -f ./switch-to-configuration-test.nix`
4. The test fails and produces the `Failed to open` warning

### Expected behavior

`switch-to-configuration` should not produce warnings.

### Additional context

It looks like `switch-to-configuration` currently can't handle service file names containing a `\`.

```
wireguard-wg0-peer-Ks9yRJIi-0vYgRmn14mIOQRwkcUGBujYINbMpik2SBIx3d.service  # file switch-to-configuration is accessing
wireguard-wg0-peer-Ks9yRJIi-0vYgRmn14mIOQRwkcUGBujYINbMpik2SBI\x3d.service # actual service file
```

I tracked the first failing commit down to #157329.

The warning seems to not create any problems for me at the moment and all services are running.

### Notify maintainers

@dasJ

### PR DESCRIPTION
**Pull request by dasJ**
systemd needs this so special characters (like the ones in wireguard
units that appear because they are part of base64) can be escaped using
the \x syntax.

Root of the issue is that `glob()` handles the backslash internally
which is obviously not what we want here.

Also add a test case and fix some perlcritic issues in the subroutine.

###### Motivation for this change

Sorry for introducing the bug, this closes #160445

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)) → `switchTest`
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dasJ** [comment]
@GrahamcOfBorg test switchTest

I'm going to merge this rather soon because the unstable channel is currently blocked and we might get this in before it advances again :)

File: nixos/modules/system/activation/switch-to-configuration.pl [hunk 1]
```diff
@@ -173,13 +173,18 @@ sub parseSystemdIni {
...
+    my ($unit_path) = @_;
 
     # Parse the main unit and all overrides
-    my %unitData;
-    parseSystemdIni(\%unitData, $_) for glob("${unitPath}{,.d/*.conf}");
-    return %unitData;
+    my %unit_data;
+    # Replace \ with \\ so glob() still works with units that have a \ in them
+    # Valid characters in unit names are ASCII letters, digits, ":", "-", "_", ".", and "\"
+    $unit_path =~ s/\\/\\\\/gmsx;
```
Comment (stigtsp) [inline nixos/modules/system/activation/switch-to-configuration.pl:183]:
```suggestion
    $unit_path =~ s/\\/\\\\/g;
```
Minor nitpick, but I don't believe the `msx` switches are needed here.

Comment (dasJ) [reply nixos/modules/system/activation/switch-to-configuration.pl:183]: Yes you are right, none of these are needed. But:
- https://metacpan.org/pod/Perl::Critic::Policy::RegularExpressions::RequireLineBoundaryMatching
- https://metacpan.org/pod/Perl::Critic::Policy::RegularExpressions::RequireDotMatchAnything
- https://metacpan.org/pod/Perl::Critic::Policy::RegularExpressions::RequireExtendedFormatting

Should I still remove them? I don't think they cause any issues when they are present

Comment (stigtsp) [reply nixos/modules/system/activation/switch-to-configuration.pl:183]: Ah right, yeah keep them in then. Interesting reasoning, haven't been running `perlcritic` with low severity before :-)

**Review by stigtsp — APPROVED** [review]

**Pull request merged**